### PR TITLE
CAZ-1316 Align database schema and logic to a new ERD

### PR DESCRIPTION
**Pull request by radekmorytko**

### CODE REVIEW

File: src/it/java/uk/gov/caz/psr/SuccessPaymentsJourneyTestIT.java [hunk 1]
```diff
@@ -187,18 +191,43 @@ public PaymentJourneyAssertion paymentEntityStatusIsUpdatedTo(PaymentStatus stat
 
     private void verifyThatPaymentEntityExistsWithStatus(PaymentStatus status) {
       int paymentsCount = JdbcTestUtils.countRowsInTableWhere(jdbcTemplate, "payment",
-          "payment_method = '" + PaymentMethod.CREDIT_CARD.name() + "' AND " +
-              "charge_paid = " + initiatePaymentRequest.getAmount() + " AND " +
-              "status = '" + status.name() + "' AND " +
-              "caz_id = '" + initiatePaymentRequest.getCleanAirZoneId().toString() + "'");
+          "payment_method = '" + PaymentMethod.CREDIT_DEBIT_CARD.name() + "' AND " +
```
Comment (plenza) [inline src/it/java/uk/gov/caz/psr/SuccessPaymentsJourneyTestIT.java]: Could you refactor this method to be more readable?

File: src/main/java/uk/gov/caz/psr/model/Payment.java [hunk 2]
```diff
@@ -1,36 +1,134 @@
...
+   * Gets the payment status provided that statuses of all entrant payments are the same (they can
+   * differ if, for example, a local authority changes the status of one vehicle entrant payment).
+   * If they are not, {@link IllegalStateException} is thrown.
+   *
+   * @return {@link PaymentStatus} for the payment if statuses of all entrant payments are the same.
+   * @throws IllegalStateException if there is at least one vehicle entrant payment's status
+   *     which is different from the others.
+   */
+  public PaymentStatus getStatus() {
+    Preconditions.checkState(vehicleEntrantPaymentsHaveOneStatus(vehicleEntrantPayments),
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/model/Payment.java]: HaveSameStatus?

File: src/main/java/uk/gov/caz/psr/model/Payment.java [hunk 3]
```diff
@@ -1,36 +1,134 @@
...
+     * @return {@link PaymentBuilder}.
+     * @throws IllegalStateException if there is at least one vehicle entrant payment's status
+     *     which is different from the others.
+     */
+    public PaymentBuilder status(PaymentStatus status) {
+      Preconditions.checkState(vehicleEntrantPaymentsHaveOneStatus(vehicleEntrantPayments),
+          "Vehicle entrant payments are empty or do not have one common status");
+      if (status == PaymentStatus.SUCCESS) {
+        authorisedTimestamp(LocalDateTime.now());
+        log.info("Setting status to '{}', setting authorised timestamp to '{}'",
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/model/Payment.java]: do you have paymentId here, so it could be logged as well?

File: src/main/java/uk/gov/caz/psr/model/Payment.java [hunk 4]
```diff
@@ -1,36 +1,134 @@
...
+
+    /**
+     * Creates a new instance of {@link VehicleEntrantPayment} with all attributes from the passed
+     * {@code vehicleEntrantPayment} object and the current value of {@link PaymentBuilder#status}.
+     *
+     * @param vehicleEntrantPayment An instance of {@link VehicleEntrantPayment} which will be
+     *     copied with a new value for {@code status}.
+     * @return A new instance of {@link VehicleEntrantPayment}.
+     */
+    private VehicleEntrantPayment updateWithCurrentStatus(
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/model/Payment.java]: VehicleEntrantPayment is "Entity" in our domain. Are we sure we want to create duplicated object (with the same id attributes, so in reality the same Entity)?

File: src/main/java/uk/gov/caz/psr/model/PaymentMethod.java [hunk 5]
```diff
@@ -1,5 +1,6 @@
 package uk.gov.caz.psr.model;
 
 public enum PaymentMethod {
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/model/PaymentMethod.java]: Javadocs

File: src/main/java/uk/gov/caz/psr/service/VehicleEntrantPaymentChargeCalculator.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+   * Calculates the charge per day based on the total amount paid and the number of days against
+   * which the payment is made.
+   *
+   * @param total The total amount to be paid.
+   * @param numberOfDays The number of days against which the payment is made.
+   * @return The charge per day.
+   */
+  public int calculateCharge(int total, int numberOfDays) {
+    int chargePerDay = total / numberOfDays;
+    if (total % numberOfDays != 0) {
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/service/VehicleEntrantPaymentChargeCalculator.java]: I think we should treat this case as serious error. So:
1. Here - add precondition to check if total is composed of numberOfDays without fraction. Throw exception and break request processing, produce 500 with critical ex.
2. In controller add validation, as soon as possible to check the same condition. If fails, return 400 to the caller with nice error message stating what happened (what they did wrong really)

File: src/test/java/uk/gov/caz/psr/util/TestObjectFactory.java [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+      UUID paymentId = UUID.randomUUID();
+      String externalId = randomExternalId();
+      return forRandomDays(paymentId, externalId);
+    }
+
+    public static Payment forRandomDays() {
+      return forRandomDays(null, null);
+    }
+
+    public static Payment forRandomDays(UUID paymentId) {
```
Comment (plenza) [inline src/test/java/uk/gov/caz/psr/util/TestObjectFactory.java]: forRandomDaysWithId

File: src/test/java/uk/gov/caz/psr/service/GetAndUpdatePaymentsServiceTest.java [hunk 8]
```diff
@@ -113,6 +115,24 @@ public void shouldUpdatePaymentStatusIfExternalPaymentIsFound() {
     verify(internalPaymentsRepository).update(internalPaymentWithExternalStatus);
   }
 
+  @Test
+  public void shouldNotUpdatePaymentIfExternalPaymentIsFoundAndItsStatusMatchesInternalOne() {
+    // given
+    UUID paymentId = UUID.fromString("744b3067-11aa-463f-b5d7-992196f959c6");
+    String externalId = "external-id";
+    PaymentStatus externalStatus = PaymentStatus.INITIATED;
+    Payment payment = mockInternalPaymentWith(paymentId, externalId);
```
Comment (plenza) [inline src/test/java/uk/gov/caz/psr/service/GetAndUpdatePaymentsServiceTest.java]:
```
Payment payment = new Payment(....);
mockPaymentRepositoryToReturn(payment);
```
I think it will be more clear for the reader - we are mocking repository to return this payment, with this attributes.

File: src/main/java/uk/gov/caz/psr/model/Payment.java (hunk 3 above)
Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/model/Payment.java]: I can log it here, however, it can be `null`.

File: src/main/java/uk/gov/caz/psr/model/Payment.java (hunk 4 above)
Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/model/Payment.java]: I'm not sure we want to do this, but I cannot think now of a better solution. I can make `VehicleEntrantPayment` mutable, but I would really rather not, taking into consideration all benefits of immutability.

File: src/main/java/uk/gov/caz/psr/service/VehicleEntrantPaymentChargeCalculator.java (hunk 6 above)
Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/service/VehicleEntrantPaymentChargeCalculator.java]: Agree, done.

Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/service/VehicleEntrantPaymentChargeCalculator.java]: Currently the returning error message to the API caller is broken, I will fix it in another story.

File: src/main/java/uk/gov/caz/psr/service/FinalizePaymentService.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+import org.springframework.stereotype.Service;
+import uk.gov.caz.psr.model.Payment;
+import uk.gov.caz.psr.model.PaymentStatus;
+import uk.gov.caz.psr.model.VehicleEntrance;
+import uk.gov.caz.psr.model.VehicleEntrantPayment;
+import uk.gov.caz.psr.repository.VehicleEntranceRepository;
+
+@Service
+@AllArgsConstructor
+public class FinalizePaymentService {
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/service/FinalizePaymentService.java]: Javadoc

File: src/main/java/uk/gov/caz/psr/service/GetAndUpdatePaymentsService.java [hunk 10]
```diff
@@ -40,29 +41,37 @@
...
-        .map(updateExternalPaymentWithInternalId(internalPayment.getId()))
+        .map(getPaymentResult -> updateInternalPaymentWith(internalPayment, getPaymentResult))
         .orElseThrow(() -> new IllegalStateException("External payment not found whereas the "
             + "internal one with id '" + id + "' and external id '" + externalPaymentId + "' "
             + "exists"));
 
-    log.info("Found the external payment, updating its status to '{}' in the database",
-        externalPayment.getStatus());
-    internalPaymentsRepository.update(externalPayment);
+    if (internalPayment.getStatus() == externalPayment.getStatus()) {
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/service/GetAndUpdatePaymentsService.java]: Method is too long - please extract bits to smaller helpers

File: src/test/java/uk/gov/caz/psr/dto/validation/constraint/AmountDivisibleByNumberOfDaysValidatorTest.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+    // given
+    InitiatePaymentRequest request = baseRequestBuilder().amount(amount).build();
+
+    // when
+    boolean result = validator.isValid(request, null);
+
+    // then
+    assertThat(result).isFalse();
+  }
+
```
Comment (plenza) [inline src/test/java/uk/gov/caz/psr/dto/validation/constraint/AmountDivisibleByNumberOfDaysValidatorTest.java]: Drop the line

Comment (radekmorytko) [reply src/test/java/uk/gov/caz/psr/dto/validation/constraint/AmountDivisibleByNumberOfDaysValidatorTest.java]: Dropped!

File: src/main/java/uk/gov/caz/psr/service/FinalizePaymentService.java (hunk 9 above)
Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/service/FinalizePaymentService.java]: Done

File: src/main/java/uk/gov/caz/psr/service/GetAndUpdatePaymentsService.java (hunk 10 above)
Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/service/GetAndUpdatePaymentsService.java]: Done

File: src/main/java/uk/gov/caz/psr/model/ExternalPaymentStatus.java [hunk 12]
```diff
@@ -1,6 +1,6 @@
 package uk.gov.caz.psr.model;
 
-public enum PaymentStatus {
+public enum ExternalPaymentStatus {
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/model/ExternalPaymentStatus.java]: Missin Javadocs

File: src/main/java/uk/gov/caz/psr/model/Payment.java [hunk 13]
```diff
@@ -1,36 +1,105 @@
...
+     * Builds the payment object.
+     *
+     * @return An instance of {@link Payment}.
+     */
+    public Payment build() {
+      Preconditions.checkState(externalStatusMatchesExternalPaymentId(),
+          "Illegal values of external payment status and ext id: (%s, %s)",
+          externalPaymentStatus, externalId);
+      Preconditions.checkState(
+          authorisedTimestamp == null ^ externalPaymentStatus == ExternalPaymentStatus.SUCCESS,
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/model/Payment.java]: This is "clever" but will hold anyone that will need to read or modify this in the future. At least add parentheses around higher priority == statements

File: src/main/java/uk/gov/caz/psr/repository/PaymentRepository.java [hunk 14]
```diff
@@ -3,80 +3,124 @@
...
+   * same status of vehicle entrant payments) into database.
    *
    * @param payment An entity object which is supposed to be saved in the database.
    * @return An instance of {@link Payment} with its internal identifier set.
-   * @throws NullPointerException if {@code payment} or {@link Payment#getCorrelationId()} is null
+   * @throws NullPointerException if {@code payment} is null
    * @throws IllegalArgumentException if {@link Payment#getId()} is not null
    */
-  public Payment insert(Payment payment) {
+  public Payment insertExternal(Payment payment) {
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/repository/PaymentRepository.java]: But this is not "external" - it is our own `Payment` - created and maintained by us. Why did you rename this method?

File: src/main/java/uk/gov/caz/psr/repository/VehicleEntrantPaymentRepository.java [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+   * Inserts the passed {@code vehicleEntrantPayments} into the database.
+   *
+   * @param vehicleEntrantPayments A list of {@link VehicleEntrantPayment} instances.
+   * @return A list of {@link VehicleEntrantPayment} with their internal identifiers set.
+   * @throws NullPointerException     if {@code vehicleEntrantPayments} is null.
+   * @throws IllegalArgumentException if {@code vehicleEntrantPayments} is empty.
+   * @throws IllegalArgumentException if {@code vehicleEntrantPayments} contains at least one object
+   *                                  whose payment id is null.
+   */
+  List<VehicleEntrantPayment> insert(List<VehicleEntrantPayment> vehicleEntrantPayments) {
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/repository/VehicleEntrantPaymentRepository.java]: I think we should mark in with `@Transactional` - if transaction is already ongoing it will just use/connect to it. If transaction is not started it will start it. If any op fails, it will rollback all inserts.

File: src/main/java/uk/gov/caz/psr/service/FinalizeVehicleEntrantService.java [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+import java.util.Optional;
+import lombok.AllArgsConstructor;
+import org.springframework.stereotype.Service;
+import uk.gov.caz.psr.model.VehicleEntrant;
+import uk.gov.caz.psr.model.VehicleEntrantPayment;
+import uk.gov.caz.psr.repository.VehicleEntrantPaymentRepository;
+
+@Service
+@AllArgsConstructor
+public class FinalizeVehicleEntrantService {
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/service/FinalizeVehicleEntrantService.java]: Javadocs

File: src/main/java/uk/gov/caz/psr/service/VehicleEntrantService.java [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+
+  /**
+   * Inserts {@code vehicleEntrant} into database (unless it exists) and connects it with the
+   * payment information provided it is valid.
+   *
+   * @param vehicleEntrant A record which is to be inserted into the database.
+   * @throws NullPointerException     if {@code vehicleEntrant} is {@code null}
+   * @throws IllegalArgumentException if {@code vehicleEntrant#id} is {@code null}
+   */
+  public void registerVehicleEntrant(VehicleEntrant vehicleEntrant) {
```
Comment (plenza) [inline src/main/java/uk/gov/caz/psr/service/VehicleEntrantService.java]: Please add `@Transactional`

File: pom.xml [hunk 18]
```diff
@@ -26,20 +26,39 @@
 
     <sonar.coverage.exclusions>
       **/configuration/**/*.java,
+      **/psr/Application.java,
       **/amazonaws/StreamLambdaHandler.java,
       **/amazonaws/CleanupDanglingPaymentsHandler.java,
-      **/psr/dto/**/*.java,
+      **/psr/dto/*.java,
```
[the remark this answers is not in the document]
Comment (radekmorytko) [reply pom.xml]: Initially all files from `dto` package were ignored, because we were working on tasks to create the swagger documentation for charge settlement API. So the exclusion is a leftover from the initial work. Good point of catching it.
I added missing unit tests for `uk.gov.caz.psr.dto.SuccessFailurePaymentStatus#from` (static factory method) and `uk.gov.caz.psr.dto.external.GetPaymentResult#getPaymentStatus` (a method which converts a status from `dto` to the one in model).
The exclusion is now only for `uk.gov.caz.psr.dto.ChargeSettlementPaymentStatus`, because no logic is coded for charge settlement.

File: src/it/java/uk/gov/caz/psr/SuccessAddingVehicleEntrantIT.java [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+  private JdbcTemplate jdbcTemplate;
+
+  private ClientAndServer mockServer;
+
+  @BeforeEach
+  public void startMockServer() {
+    mockServer = startClientAndServer(1080);
+    RestAssured.port = randomServerPort;
+    RestAssured.baseURI = "http://localhost";
+    RestAssured.basePath = "/v1/payments/vehicle-entrants";
```
[the remark this answers is not in the document]
Comment (radekmorytko) [reply src/it/java/uk/gov/caz/psr/SuccessAddingVehicleEntrantIT.java]: Fixed

File: src/it/java/uk/gov/caz/psr/SuccessAddingVehicleEntrantIT.java [hunk 20]
```diff
@@ -0,0 +1,177 @@
...
+          .accept(MediaType.APPLICATION_JSON.toString())
+          .contentType(MediaType.APPLICATION_JSON.toString())
+          .header(Constants.X_CORRELATION_ID_HEADER, correlationId)
+          .body(toJsonString(vehicleEntrantRequest))
+          .when()
+          .post()
+          .then()
+          .header(Constants.X_CORRELATION_ID_HEADER, correlationId)
+          .statusCode(HttpStatus.OK.value())
+          .extract()
```
[the remark this answers is not in the document]
Comment (radekmorytko) [reply src/it/java/uk/gov/caz/psr/SuccessAddingVehicleEntrantIT.java]: I removed the extraction bits as currently the controller returns `200` status code with no body.

File: src/it/java/uk/gov/caz/psr/repository/PaymentRepositoryTestIT.java [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+
+  @Autowired
+  private PaymentRepository paymentRepository;
+
+  @Sql(scripts = {"classpath:data/sql/clear-all-payments.sql",
+      "classpath:data/sql/clear-all-vehicle-entrants.sql",
+      "classpath:data/sql/add-payment-with-null-external-id.sql"},
+      executionPhase = ExecutionPhase.BEFORE_TEST_METHOD)
+  @Sql(scripts = {"classpath:data/sql/clear-all-payments.sql",
+      "classpath:data/sql/clear-all-vehicle-entrants.sql"},
```
[the remark this answers is not in the document]
Comment (radekmorytko) [reply src/it/java/uk/gov/caz/psr/repository/PaymentRepositoryTestIT.java]: They're not, thanks!

File: src/main/java/uk/gov/caz/psr/model/ExternalPaymentStatus.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/model/ExternalPaymentStatus.java]: Added.

File: src/main/java/uk/gov/caz/psr/repository/PaymentRepository.java (hunk 14 above)
Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/repository/PaymentRepository.java]: The reason is that there are two types of payments based on _where_ the payment takes place:
- the ones conducted by GOV UK Pay (_external_)
- the ones managed by LAs (_internal_ - this may not be the best name, though)

`insertExternal` contains '`payment.getExternalPaymentStatus()` cannot be `null`' precondition, which is not needed when it comes to payments done in LAs (we will handle those when charge-settlement API is in place)

I think I will rename this method to `insertWithExternalStatus` to indicate the intention more clearly.

File: src/main/java/uk/gov/caz/psr/repository/VehicleEntrantPaymentRepository.java (hunk 15 above)
Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/repository/VehicleEntrantPaymentRepository.java]: Good point, added.

Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/repository/VehicleEntrantPaymentRepository.java]: I also annotated `uk.gov.caz.psr.repository.VehicleEntrantPaymentRepository#update(java.util.List<uk.gov.caz.psr.model.VehicleEntrantPayment>)` with `@Transactional`.

File: src/main/java/uk/gov/caz/psr/service/FinalizeVehicleEntrantService.java (hunk 16 above)
Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/service/FinalizeVehicleEntrantService.java]: Done.

File: src/main/java/uk/gov/caz/psr/service/VehicleEntrantService.java (hunk 17 above)
Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/service/VehicleEntrantService.java]: Done.

File: src/main/java/uk/gov/caz/psr/model/Payment.java (hunk 13 above)
Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/model/Payment.java]: I somehow missed your comment when doing changes, will add it later in a separate PR.

File: src/main/java/uk/gov/caz/psr/model/ExternalPaymentStatus.java (hunk 12 above)
Comment (radekmorytko) [reply src/main/java/uk/gov/caz/psr/model/ExternalPaymentStatus.java]: Done
